services/src/main/java/org/keycloak/vault/FilesKeystoreVaultProvider.java

```diff
@@ -0,0 +1,84 @@
+package org.keycloak.vault;
+
+import java.nio.ByteBuffer;
+import java.security.*;
+import java.security.spec.InvalidKeySpecException;
+import org.jboss.logging.Logger;
+
+import javax.annotation.Nonnull;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.cert.CertificateException;
+import java.util.List;
+import java.util.Optional;
+
+public class FilesKeystoreVaultProvider extends AbstractVaultProvider {
+
+    private static final Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final Path keystorePath;
+    private final String keystorePass;
+    private final String keystoreType;
+    private static final String encryptionAlgorithm = "PBE";
+
+    /**
+     * Creates a new {@link FilesKeystoreVaultProvider}.
+     *
+     * @param keystorePath A path to a vault. Can not be null.
+     * @param keystorePass A password to a vault. Can not be null.
+     * @param realmName A realm name. Can not be null.
+     */
+    public FilesKeystoreVaultProvider(@Nonnull Path keystorePath, @Nonnull String keystorePass, @Nonnull String realmName, @Nonnull List<VaultKeyResolver> resolvers) {
+        super(realmName, resolvers);
+        this.keystorePath = keystorePath;
+        this.keystorePass = keystorePass;
+        this.keystoreType = "PKCS12";
+        logger.debugf("KeystoreVaultProvider will operate in %s directory", keystorePath.toAbsolutePath());
+    }
+
+    /**
+     * Creates a new {@link FilesKeystoreVaultProvider}.
+     *
+     * @param keystoreType Specifies a type of keystore. Can not be null. Default value is PKCS12.
+     */
+    public FilesKeystoreVaultProvider(@Nonnull Path keystorePath, @Nonnull String keystorePass, @Nonnull String keystoreType,
+                                      @Nonnull String realmName, @Nonnull List<VaultKeyResolver> resolvers) {
+        super(realmName, resolvers);
+        this.keystorePath = keystorePath;
+        this.keystorePass = keystorePass;
+        this.keystoreType = keystoreType;
+        logger.debugf("KeystoreVaultProvider will operate in %s directory", keystorePath.toAbsolutePath());
+    }
+
+    @Override
+    protected VaultRawSecret obtainSecretInternal(String alias) {
+        if (!Files.exists(keystorePath)) {
+            logger.warnf("Cannot find secret alias %s in %s", alias, keystorePath);
+            return DefaultVaultRawSecret.forBuffer(Optional.empty());
+        }
+        KeyStore ks;
+        SecretKeyFactory factory;
+        PBEKeySpec keySpec;
+        try {
+            ks = KeyStore.getInstance(keystoreType);
+            ks.load(Files.newInputStream(keystorePath), keystorePass.toCharArray());
```

Comment:
@Pepo48 it would be nice to mitigate _Path Traversal_ before we merge those changes by escaping or validating the user input when _keystorePath_ is provided. Otherwise, we may introduce a new vulnerability.

Please see:
- https://owasp.org/www-community/attacks/Path_Traversal
- https://codeql.github.com/codeql-query-help/java/java-path-injection/